### ISSUE TITLE
MacOS / OSX GPU support

### ISSUE THREAD

**Issue by JobJob** [malmaud/TensorFlow.jl#103]
Hi, just started taking a look at this package, looks nice, great work!

I'm on Mac (Sierra 10.12) and have managed to seemingly get TensorFlow.jl running on my GPU.

Basically, I just followed the instructions [here](https://www.tensorflow.org/versions/r0.11/get_started/os_setup.html#installing-from-sources) and ran:
`bazel build -c opt --config=cuda //tensorflow:libtensorflow.so`
and copied the resulting .so file to 
`~/.julia/v0.5/TensorFlow/deps/usr/bin/libtensorflow_c.dylib`.

So, am I able to upload my pre-built `libtensorflow_c.dylib` somewhere so others can use it? Will that work on other machines or do I need some other build flags (e.g. related to paths to python and cuda)?

If it will work:
What GPU compute capabilities should I compile with?
Any other build flags or ./configure options? I set Google Cloud Platform off and HDFS off and opencl off

Is there anything to test other than `julia ~/.julia/v0.5/TensorFlow/test/runtests.jl` (which had no errors btw).

Cheers,
Joel

---
P.S. This is how I was testing whether it used the GPU:
```
using TensorFlow
const tf = TensorFlow.tensorflow
sess = TensorFlow.Session(config=tf.ConfigProto(log_device_placement=true))
x = rand(100,100)
y = rand(100,100)

a = TensorFlow.constant(x)
b = TensorFlow.Variable(y)

c = a*b
run(sess, TensorFlow.initialize_all_variables())
res = run(sess, c)
Base.Test.@test res ≈ x*y
```
despite the fact that, even when using the default .dylib downloaded by build.jl, it shows:
```
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcublas.dylib locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcudnn.dylib locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcufft.dylib locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcuda.1.dylib locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcurand.dylib locally
```
The next section shows that it's running on the CPU:
```
Device mapping: no known devices.
node: /job:localhost/replica:0/task:0/cpu:0
node/Assign: /job:localhost/replica:0/task:0/cpu:0
NoOp: /job:localhost/replica:0/task:0/cpu:0
MatMul: /job:localhost/replica:0/task:0/cpu:0
Const_2: /job:localhost/replica:0/task:0/cpu:0
Const: /job:localhost/replica:0/task:0/cpu:0
```
With my updated dylib it shows it's using the GPU.

**Comment by JobJob** [comment]
Also, forgot to mention, for anyone else trying to compile a tensorflow .dylib/.so with GPU support on Mac, as noted on the CUDA download page: Xcode (clang) 8 is not currently supported with the CUDA Toolkit 8 compiler for Mac, so you need to use Xcode (clang) 7.x with the CUDA Toolkit 8.

I had to download and install the Xcode command line developer tools 7.3 (for OSX 10.11) from [here](https://developer.apple.com/download/more/), and run 
```sudo xcode-select -s /Library/Developer/CommandLineTools```

to switch back to your up-to-date Xcode after you've compiled tensorflow, use
`sudo xcode-select -s /Applications/Xcode.app/Contents/Developer`

**Comment by malmaud** [comment]
Great work, thanks! I think I better compile the final uploaded binary myself so I can update it as upstream TensorFlow updates, but your tip about `xcode-select` is really helpful.

**Comment by JobJob** [comment]
Yep, sounds good, just figured it's possible you might not have access to a mac.

Also FYI, for some reason HomeBrew.jl's Brew started complaining again about outdated XCode yesterday (maybe I'd only used Macports after the `xcode-select` back to the Xcode.app path), so I was forced to do this:
```
sudo mv /Library/Developer/CommandLineTools/ /Library/Developer/CommandLineTools7.3/
sudo ln -s /Applications/Xcode.app/Contents/Developer /Library/Developer/CommandLineTools
```
which Brew was satisfied with.

**Comment by JobJob** [comment]
Another thing I just remembered, I found that I had to set my LD_LIBRARY_PATH environment variable in my `~/.profile` file to include the cuda install path: `/usr/local/cuda/lib` (by default) to get TensorFlow to load correctly.

Furthermore, when running in Atom / Juno, Atom doesn't inherit the LD_LIBRARY_PATH environment variable successfully and as a result `using Tensorflow` causes a segmentation fault. The simplest solution that worked for me is I added:
```
process.env.LD_LIBRARY_PATH = "/usr/local/cuda/lib:path/to/other/lib/dir1:..."
```
to my init.coffee

This second one took me a while to track down, so hope it helps someone.
